components/tidb_query/src/codec/row/v2/encoder.rs

```diff
@@ -0,0 +1,295 @@
+// number of non-null values: 2 bytes
+// number of null values: 2 bytes
+// column ids:  ids of non-null values + ids of null values, when flag == 1 (big), id is 4 bytes, otherwise 1 byte
+// non-null values offset: when big, offset is 4 bytes, otherwise 2 bytes
+pub fn encode(datums: Vec<Datum>, ids: &[i64]) -> Result<Vec<u8>> {
+    let row = Row::new_row(datums, ids);
+    let mut wtr = Vec::with_capacity(row.approx_size());
+    wtr.write_row(row)?;
+    Ok(wtr)
+}
+
+trait RowEncoder: NumberEncoder {
+    fn write_row(&mut self, row: Row) -> Result<()> {
+        let mut offset_wtr = Vec::with_capacity(row.offsets_len());
+        let mut value_wtr = Vec::with_capacity(row.value_approx_size);
+
+        self.write_u8(super::CODEC_VERSION)?;
+        self.write_flag(row.is_big)?;
+        self.write_u16_le(row.non_null_cols.len() as u16)?;
+        self.write_u16_le(row.null_cols.len() as u16)?;
+
+        for (datum, id) in row.non_null_cols {
+            self.write_id(row.is_big, id)?;
+            value_wtr.write_datum(datum)?;
+            offset_wtr.write_offset(row.is_big, value_wtr.len())?;
+        }
+        for (_, id) in row.null_cols {
+            self.write_id(row.is_big, id)?;
+        }
+        self.write_bytes(&offset_wtr)?;
+        self.write_bytes(&value_wtr)?;
+        Ok(())
+    }
+
+    #[inline]
+    fn write_flag(&mut self, is_big: bool) -> codec::Result<()> {
+        self.write_u8(if is_big { 1 } else { 0 })
+    }
+
+    #[inline]
+    fn write_id(&mut self, is_big: bool, id: i64) -> codec::Result<()> {
+        if is_big {
+            self.write_u32_le(id as u32)
+        } else {
+            self.write_u8(id as u8)
+        }
+    }
+
+    #[inline]
+    fn write_offset(&mut self, is_big: bool, offset: usize) -> codec::Result<()> {
+        if is_big {
+            self.write_u32_le(offset as u32)
+        } else {
+            self.write_u16_le(offset as u16)
+        }
+    }
+}
+impl RowEncoder for Vec<u8> {}
+
+trait DatumEncoder: NumberEncoder + DecimalEncoder + JsonEncoder {
+    #[inline]
+    fn write_datum(&mut self, datum: Datum) -> Result<()> {
+        match datum {
+            Datum::I64(i) => self.encode_i64(i).map_err(Error::from),
+            Datum::U64(u) => self.encode_u64(u).map_err(Error::from),
+            Datum::Bytes(b) => self.write_bytes(&b).map_err(Error::from),
+            Datum::Dec(d) => {
+                let (prec, frac) = d.prec_and_frac();
+                self.write_decimal(&d, prec, frac)?;
+                Ok(())
+            }
+            Datum::Json(j) => self.write_json(&j),
+            Datum::F64(_) | Datum::Time(_) | Datum::Dur(_) | Datum::Null => unreachable!(),
+            Datum::Min | Datum::Max => unimplemented!("not supported yet"),
+        }
+    }
+
+    #[allow(clippy::match_overlapping_arm)]
+    #[inline]
+    fn encode_i64(&mut self, v: i64) -> codec::Result<()> {
+        match v {
```

Comment:
I think we could use a generic function call `is_bounded_by` to extract this kind of code. It is worth to have a try.